feat: add ability to overload all cyberware slots

13k opened the pull request:
## What it does

Adds ability to overload all cyberware slots (except Ocular System) with configuration managed by [Mod Settings](https://github.com/jackhumbert/mod_settings).

## Motivation

There's a couple mods in nexusmods that patch System-Ex to overload other cyberware slots. [Hands-Ex](https://www.nexusmods.com/cyberpunk2077/mods/4718) hardcodes hands slots to 2 and [All Cyberware-EX](https://www.nexusmods.com/cyberpunk2077/mods/5097) hardcodes all slots to the maximum (?) (not sure if there's a maximum).

First, these mods cover extremely specific cases and if users want a different set of options, these mods must either add settings (on top of manually patching System-Ex), or create new mods with different hardcoded values.

Secondly, this manual patching on other mods is unmaintainable. When System-Ex is updated, these patches would most likely be broken, having to merge System-Ex's updates to their code.

So, to integrate everything, I think the best solution is to modify System-Ex with the ability to configure all slots.

## Notes

* All slot count settings default to vanilla values, except System Replacement, where System-Ex already set it to 2.
* I'm not sure how to implement an `IScriptable` that can be used as a singleton that is automatically initialized by the game, so I used a `ScriptableSystem` with `OnAttach()`, following the same pattern as `Codeware.Registry.RegistrySystem`
* I don't know if any cyberware needs special-case handling like System Replacement. If so, then this PR needs to be updated with it
* Ocular system is hardcoded to 1, since it doesn't make sense to use multiple
* Adding a dependency (Mod Settings) may be a deterrent to get this merged, so I understand if it's rejected
* I did not change the mod version. Since it adds a dependency, which is a breaking change, it's up to you if you want to bump the major version. (It seems the mod will still work without Mod Settings installed and will simply use the default values)

13k committed:
feat: add ability to overload all cyberware slots

Adds ability to overload all cyberware slots (except Ocular System) with configuration managed by
Mod Settings.

All slot count settings default to vanilla values, except Operating System, where System-Ex already
set it to 2.

psiberx commented:
Hi. Thanks for the PR. You made a good points. As for details:
- `ScriptableSystem` is the way to go.
- Yeah, the good thing about Mod Settings is that it can be an optional requirement. The mod will work using default settings without it.
- Active cyberware may require special handling. I can think of a legs slot (double jump, charged jump, hovering). I'd check and handle this as the mod's idea is to provide fully functional extensions.

13k commented:
Yeah, I need to check cyberwares for special cases. I personally don't overload other cyberwares other than System Replacement and Hands (System-Ex + Hands-Ex). I'll also check All Cyberware-Ex to see if they handle special cases.

13k committed:
fix: change `SystemEx` lifecycle and a couple refactorings

For some reason `ScriptableSystem.OnAttach()` breaks CET's
`Game.GetPlayer()`. I have no idea why. There are other mods/libraries
using `OnAttach()` without issues. Doesn't seem related to
`EquipmentSystemPlayerData` requesting an instance too soon, since
`OnAttach` seems to be called earlier (after CET loads the Lua modules
and before triggering "onInit").

My "solution" is a workaround to manually call the `Initialize()`
method and change it to not re-initialize every time. This is garbage.
If other places used `SystemEx` instance, they would also need to
manually `Initialize()` it.

A couple refactorings:

* Move slots logic from `SystemEx` to `SlotsConfig`
* Rename the `OverloadCWSlots(SystemEx, gamedataEquipmentArea)` overload
  to `OverloadCWSlotType(SystemEx, gamedataEquipmentArea)`

13k commented:
I'm closing this since 1.0.5 adds Hands cyberware option and OS+Hands is everything I use.
